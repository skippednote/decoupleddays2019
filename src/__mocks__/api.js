import {
  sessions,
  categories,
  users,
  createSessionSuccess,
  deleteSessionSuccess
} from "../utils/fakes";

export const API = {
  getSessions: jest.fn(url => {
    if (url) {
      return sessions[0];
    } else {
      return sessions;
    }
  }),
  createSession: jest.fn(() => {
    return createSessionSuccess;
  }),
  updateSession: jest.fn(() => {
    return createSessionSuccess;
  }),
  deleteSession: jest.fn(() => {
    return deleteSessionSuccess;
  }),
  getCategories: jest.fn(() => {
    return categories;
  }),
  getUsers: jest.fn(() => {
    return users;
  })
};
