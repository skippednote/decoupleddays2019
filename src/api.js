import { useState, useEffect } from "react";
import axios from "axios";
import normalize from "json-api-normalize";
import Axios from "axios";
import * as Sentry from "@sentry/browser";

class APIFactory {
  constructor() {
    this.url = "";
    this.fields = null;
    this.defaultFields = [
      "id",
      "title",
      "created",
      "field_room",
      "field_date",
      "field_length",
      "field_speakers",
      "field_speakers.id",
      "field_speakers.user_picture.uri",
      "field_speakers.field_full_name",
      "field_speakers.field_company",
      "field_category.id",
      "field_category.name"
    ];
    this.baseURL = `http://sessions.lndo.site/api`;
  }

  getSessions(url = "", fields = []) {
    const [session, setSession] = useState(null);
    useEffect(() => {
      const cancelToken = axios.CancelToken;
      const source = cancelToken.source();
      try {
        const getS = async () => {
          const { data: response } = await axios.get(
            this._generateURL("node/session/" + url, true),
            {
              cancelToken: source.token
            }
          );
          let f = [...this.defaultFields];
          if (fields) {
            f = [...f, ...fields];
          }
          const normalizedData = normalize(response).get(f);
          setSession(normalizedData);
        };
        getS();
      } catch (error) {
        setSession("error");
        Sentry.withScope(scope => {
          scope.setExtras(error.message);
          Sentry.captureException(error);
        });
      }

      return () => source.cancel();
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return session;
  }

  getCategories() {
    const [categories, setCategories] = useState(null);

    useEffect(() => {
      const cancelToken = axios.CancelToken;
      const source = cancelToken.source();
      try {
        const get = async () => {
          const { data } = await axios.get(
            this._generateURL("taxonomy_term/category")
          );

          setCategories(
            data.data.map(category => ({
              value: category.id,
              label: category.attributes.name
            }))
          );
        };

        get();
      } catch (error) {
        setCategories("error");
        Sentry.withScope(scope => {
          scope.setExtras(error.message);
          Sentry.captureException(error);
        });
      }

      return () => source.cancel();
    }, []);

    return categories;
  }

  getUsers() {
    const [users, setUsers] = useState(null);
    useEffect(() => {
      const cancelToken = axios.CancelToken;
      const source = cancelToken.source();
      try {
        const get = async () => {
          const { data } = await axios.get(this._generateURL("user/user"));
          setUsers(
            data.data.map(u => ({
              value: u.id,
              label: u.attributes.field_full_name || "Anonymous"
            }))
          );
        };

        get();
      } catch (error) {
        setUsers("error");
        Sentry.withScope(scope => {
          scope.setExtras(error.message);
          Sentry.captureException(error);
        });
      }

      return () => source.cancel();
    }, []);

    return users;
  }

  async createSession(values) {
    try {
      const data = this._generateSubmitData(values);
      const { data: session } = await axios.post(
        this._generateURL("node/session"),
        data,
        {
          headers: {
            Accept: "application/vnd.api+json",
            "Content-Type": "application/vnd.api+json",
            Authorization: window.localStorage.getItem("token")
          }
        }
      );
      return { id: session.data.id, error: false };
    } catch (error) {
      Sentry.withScope(scope => {
        scope.setExtras(error.message);
        Sentry.captureException(error);
      });
      return { id: false, error: true };
    }
  }

  async updateSession(values, id) {
    try {
      const data = this._generateSubmitData(values, id);
      const { data: session } = await axios.patch(
        this._generateURL("node/session/" + id),
        data,
        {
          headers: {
            Accept: "application/vnd.api+json",
            "Content-Type": "application/vnd.api+json",
            Authorization: window.localStorage.getItem("token")
          }
        }
      );
      return { id: session.data.id, error: false };
    } catch (error) {
      Sentry.withScope(scope => {
        scope.setExtras(error.message);
        Sentry.captureException(error);
      });
      return { id: false, error: true };
    }
  }

  async deleteSession(id) {
    try {
      const token = window.localStorage.getItem("token");
      await Axios.delete(this._generateURL(`node/session/${id}`), {
        headers: {
          Authorization: token
        }
      });

      return { error: false, success: true };
    } catch (error) {
      Sentry.withScope(scope => {
        scope.setExtras(error.message);
        Sentry.captureException(error);
      });
      return { error: true, success: false };
    }
  }

  _generateURL(subpath = "", include = false) {
    let queryParams = "";
    if (include) {
      queryParams = `?include=field_speakers,field_speakers.user_picture,field_category&sort=-created`;
    }
    return `${this.baseURL}/${subpath}${queryParams}`;
  }

  _generateSubmitData(values, id) {
    const data = {
      data: {
        type: "node--session",
        attributes: {
          title: values.title,
          body: {
            value: values.body,
            format: "full_html"
          },
          field_length: Number(values.slength),
          field_room: values.room,
          field_date: `${values.date}T${values.time}+00:00`
        },
        relationships: {
          field_category: {
            data: {
              type: "taxonomy_term--category",
              id: values.category
            }
          },
          field_speakers: {
            data:
              values.speakers &&
              values.speakers.map(speaker => ({
                type: "user--user",
                id: speaker.value
              }))
          }
        }
      }
    };

    if (id) {
      data.data.id = id;
      return data;
    }

    return data;
  }
}
const API = new APIFactory();

export { API };
