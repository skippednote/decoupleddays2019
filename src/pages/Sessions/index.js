import React, { lazy, Suspense } from "react";
import { API } from "../../api";
const SessionList = lazy(() => import("../../components/SessionList"));

function Sessions() {
  const sessions = API.getSessions();

  if (sessions) {
    return (
      <Suspense fallback={<div>Loading component...</div>}>
        <SessionList sessions={sessions} />
      </Suspense>
    );
  }

  return <div data-testid="loading-data">Loading data...</div>;
}

export default Sessions;
