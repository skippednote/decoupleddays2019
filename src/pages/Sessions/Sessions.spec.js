import React from "react";
import { render, wait } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Sessions from ".";
import { API as MockAPI } from "../../api";

jest.mock("../../api.js");

afterEach(() => {
  MockAPI.getSessions.mockClear();
});

it("renders the component", async () => {
  const { container, getByTestId } = render(
    <MemoryRouter>
      <Sessions />
    </MemoryRouter>
  );

  expect(container).toMatchSnapshot();

  await wait(() => expect(getByTestId("session")).toBeInTheDocument());
  expect(container).toMatchSnapshot();
});

it("renders all the sessions", async () => {
  const { queryAllByTestId } = render(
    <MemoryRouter>
      <Sessions />
    </MemoryRouter>
  );

  await wait(() => {
    expect(queryAllByTestId("session-item").length).toEqual(4);
  });
});

it("calls the api", async () => {
  const { getByTestId } = render(
    <MemoryRouter>
      <Sessions />
    </MemoryRouter>
  );
  await wait(() => expect(getByTestId("session")).toBeInTheDocument());
  expect(MockAPI.getSessions).toHaveBeenCalledTimes(1);
});

it("renders the loading message while waiting for API response", async () => {
  MockAPI.getSessions.mockReset();
  MockAPI.getSessions.mockImplementationOnce(() => {
    setTimeout(() => {}, 100);
  });
  const { getByTestId } = render(
    <MemoryRouter>
      <Sessions />
    </MemoryRouter>
  );
  expect(getByTestId("loading-data")).toBeInTheDocument();
});
