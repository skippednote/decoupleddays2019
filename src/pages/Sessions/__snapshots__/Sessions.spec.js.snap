// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <div>
    Loading component...
  </div>
</div>
`;

exports[`renders the component 2`] = `
<div>
  <div>
    <div>
      <button
        class="button"
        data-testid="show-filters"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="14"
            rx="7"
            ry="7"
            width="22"
            x="1"
            y="5"
          />
          <circle
            cx="16"
            cy="12"
            r="3"
          />
        </svg>
         
        <div
          class="filter-label"
        >
          Hide Filters
        </div>
      </button>
    </div>
    <div
      data-testid="session"
    >
      <section
        class="session-item"
        data-testid="session-item"
      >
        <h4
          class="session-title"
          data-testid="session-title"
        >
          <a
            href="/017a6b73-fdcb-473f-aa49-8ae0073ab076"
          >
            a
          </a>
          <span
            class="session-length"
          >
            25
             minutes
          </span>
        </h4>
        <div
          class="session-meta"
        >
          <p
            class="session-room"
            data-testid="session-room"
          >
            room1
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-date"
            data-testid="session-date"
          >
            2019-07-09T00:01:00+00:00
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-category"
            data-testid="session-category"
          >
            Decoupled Drupal
          </p>
        </div>
        <ul
          class="session-speakers"
          data-testid="session-speakers"
        >
          <li>
            <p>
              Bassam Ismail
               
              (Axelerant)
            </p>
          </li>
        </ul>
      </section>
      <section
        class="session-item"
        data-testid="session-item"
      >
        <h4
          class="session-title"
          data-testid="session-title"
        >
          <a
            href="/8a940125-c1ab-40f0-abfb-c45ee7a926a0"
          >
            edit
          </a>
          <span
            class="session-length"
          >
            25
             minutes
          </span>
        </h4>
        <div
          class="session-meta"
        >
          <p
            class="session-room"
            data-testid="session-room"
          >
            room2
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-date"
            data-testid="session-date"
          >
            2019-07-10T01:01:00+00:00
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-category"
            data-testid="session-category"
          >
            JavaScript and JAMstack
          </p>
        </div>
        <ul
          class="session-speakers"
          data-testid="session-speakers"
        >
          <li>
            <p>
              Bassam Ismail
               
              (Axelerant)
            </p>
          </li>
        </ul>
      </section>
      <section
        class="session-item"
        data-testid="session-item"
      >
        <h4
          class="session-title"
          data-testid="session-title"
        >
          <a
            href="/10ba43a8-f0ef-4e8c-8954-77caf2261833"
          >
            3
          </a>
          <span
            class="session-length"
          >
            50
             minutes
          </span>
        </h4>
        <div
          class="session-meta"
        >
          <p
            class="session-room"
            data-testid="session-room"
          >
            room3
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-date"
            data-testid="session-date"
          >
            3333-03-31T03:03:00+00:00
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-category"
            data-testid="session-category"
          >
            JavaScript and JAMstack
          </p>
        </div>
        <ul
          class="session-speakers"
          data-testid="session-speakers"
        >
          <li>
            <p>
               
            </p>
          </li>
        </ul>
      </section>
      <section
        class="session-item"
        data-testid="session-item"
      >
        <h4
          class="session-title"
          data-testid="session-title"
        >
          <a
            href="/8f462bbf-193e-401e-a02d-cdb4e51d0946"
          >
            Hello
          </a>
          <span
            class="session-length"
          >
            50
             minutes
          </span>
        </h4>
        <div
          class="session-meta"
        >
          <p
            class="session-room"
            data-testid="session-room"
          >
            room1
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-date"
            data-testid="session-date"
          >
            2019-07-17T01:02:00+00:00
          </p>
          <span
            class="spacer"
          >
            •
          </span>
          <p
            class="session-category"
            data-testid="session-category"
          >
            JavaScript and JAMstack
          </p>
        </div>
        <ul
          class="session-speakers"
          data-testid="session-speakers"
        >
          <li>
            <p>
              Bassam Ismail
               
              (Axelerant)
            </p>
          </li>
        </ul>
      </section>
    </div>
  </div>
</div>
`;
